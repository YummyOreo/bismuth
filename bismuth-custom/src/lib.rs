#![allow(dead_code, unused)]
use std::collections::HashMap;

use bismuth_parser::{
    custom::CustomElm,
    tree::{Ast, Element, Kind},
    Metadata, Parser,
};

mod builtin;
pub mod plugin;
pub mod template;

pub struct Custom {
    name: String,
    data: HashMap<String, String>,
    template: Option<template::Template>,
    plugin: Option<plugin::Plugin>,
}

impl Custom {
    pub fn new(name: String, data: HashMap<String, String>) -> Self {
        Custom {
            name,
            data,
            template: None,
            plugin: None,
        }
    }

    pub fn from_elm(elm: &CustomElm) -> Self {
        Self::new(elm.name.clone(), elm.values.clone())
    }

    fn find_plugin(&mut self) -> Option<plugin::Plugin> {
        // REDO THIS WHEN YOU IMPLEMENT PLUGINS
        match self.name.to_string().as_str() {
            "blog list" | "blogs" | "bloglist" => {}
            "navbar" => {}
            "footer" => {}
            _ => {}
        }
        None
    }
    fn find_template(&mut self) -> Option<template::Template> {
        // REDO THIS WHEN YOU IMPLEMENT TEMPLATE
        match self.name.to_string().as_str() {
            "blog list" | "blogs" | "bloglist" => {}
            "navbar" => {}
            "footer" => {}
            _ => {}
        }
        None
    }

    pub fn find(&mut self) {
        self.plugin = self.find_plugin();
        self.template = self.find_template();
    }
}

fn get_customs(ast: &Ast) -> Vec<&Element> {
    let elements: Vec<&Element> = ast.elements.iter().collect();
    elements
        .iter()
        .filter_map(|&e| {
            if let Kind::CustomElement(_) = &e.kind {
                Some(e)
            } else {
                None
            }
        })
        .collect::<Vec<&Element>>()
}

pub fn parse_custom(mut target: Parser, others: Vec<&Parser>) -> Parser {
    let custom_elms = get_customs(&target.ast);
    let customs: Vec<Custom> = custom_elms
        .iter()
        .filter_map(|e| {
            if let Kind::CustomElement(c) = &e.kind {
                let mut custom = Custom::from_elm(c);
                custom.find();
                Some(custom)
            } else {
                None
            }
        })
        .collect();
    todo!()
}

#[cfg(test)]
mod test_utils {
    use super::*;
    use regex::Regex;

    macro_rules! snapshot {
        ($content:tt) => {
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("../testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!($content);
            });
        };
    }
    #[test]
    fn get_customs_test() {
        let mut parser =
            bismuth_parser::Parser::new_test("/test/", "%{{\nname: test\nother: key\n}}");
        parser.parse().unwrap();

        let customs = format!("{:#?}", get_customs(&parser.ast));
        let re = Regex::new(r"id: \d+").unwrap();
        let customs = re.replace_all(&customs, "id: [redacted]").to_string();
        snapshot!(customs);
    }
}
