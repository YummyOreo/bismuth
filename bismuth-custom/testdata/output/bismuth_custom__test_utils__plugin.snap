---
source: bismuth-custom/src/lib.rs
expression: customs
---
Parser {
    lexer: Lexer {
        path: "/test/",
        chars: [
            '%',
            '{',
            '{',
            '\n',
            'n',
            'a',
            'm',
            'e',
            ':',
            ' ',
            'n',
            'a',
            'v',
            'b',
            'a',
            'r',
            '\n',
            'o',
            't',
            'h',
            'e',
            'r',
            ':',
            ' ',
            'k',
            'e',
            'y',
            '\n',
            '}',
            '}',
            '\n',
        ],
        position: 31,
        current_token: Token {
            start: 31,
            end: 31,
            kind: EndOfFile,
            text: [],
        },
        tokens: [
            Token {
                start: 0,
                end: 0,
                kind: StartOfFile,
                text: [],
            },
            Token {
                start: 0,
                end: 0,
                kind: Percent,
                text: [
                    '%',
                ],
            },
            Token {
                start: 1,
                end: 2,
                kind: CurlybraceLeft,
                text: [
                    '{',
                    '{',
                ],
            },
            Token {
                start: 3,
                end: 3,
                kind: EndOfLine,
                text: [
                    '\n',
                ],
            },
            Token {
                start: 4,
                end: 15,
                kind: Text,
                text: [
                    'n',
                    'a',
                    'm',
                    'e',
                    ':',
                    ' ',
                    'n',
                    'a',
                    'v',
                    'b',
                    'a',
                    'r',
                ],
            },
            Token {
                start: 16,
                end: 16,
                kind: EndOfLine,
                text: [
                    '\n',
                ],
            },
            Token {
                start: 17,
                end: 26,
                kind: Text,
                text: [
                    'o',
                    't',
                    'h',
                    'e',
                    'r',
                    ':',
                    ' ',
                    'k',
                    'e',
                    'y',
                ],
            },
            Token {
                start: 27,
                end: 27,
                kind: EndOfLine,
                text: [
                    '\n',
                ],
            },
            Token {
                start: 28,
                end: 29,
                kind: CurlybraceRight,
                text: [
                    '}',
                    '}',
                ],
            },
            Token {
                start: 30,
                end: 30,
                kind: EndOfLine,
                text: [
                    '\n',
                ],
            },
            Token {
                start: 31,
                end: 31,
                kind: EndOfFile,
                text: [],
            },
        ],
    },
    index: 10,
    metadata: Metadata {
        absolute_path: "/test/",
        fontmatter: FontMatter {
            title: Some(
                "test",
            ),
            path: Some(
                "/test/",
            ),
            kind: Some(
                "default",
            ),
            values: None,
        },
    },
    current_element: None,
    state: State {
        new_line: true,
        inside: [],
        indent_level: 0,
    },
    has_custom: true,
    ast: Ast {
        elements: [
            Element {
                kind: CustomElement(
                    CustomElm {
                        name: "navbar",
                        values: {
                            "other": "key",
                        },
                        body: None,
                        template: None,
                    },
                ),
                elements: [
                    Element {
                        kind: Text,
                        elements: [],
                        text: Some(
                            "other: key",
                        ),
                        attrs: {},
                        id: [redacted],
                    },
                ],
                text: None,
                attrs: {},
                id: [redacted],
            },
            Element {
                kind: EndOfLine,
                elements: [],
                text: None,
                attrs: {},
                id: [redacted],
            },
        ],
    },
}
